### ISSUE TITLE
search for taggables that match all tags

### ISSUE THREAD

**Issue by grosser** [pat/gutentag#34]
tagged_with is a OR query, since it uses IN ... how about adding `tagged_with_all` ... or `tagged_with x, all: true` ?

code I'm using atm:
```
        scope = scope.joins(:taggings).where("gutentag_taggings.tag_id" => v)
        if v.size == 1
          scope.distinct
        else
          scope.having("count(repositories.id) = #{v.size}").group('repositories.id')
        end
```

**Comment by pat** [comment]
Currently unsure if I want to add this in.

Like I've discussed in #32, I feel like providing many ways to search by tags is getting closer to what specific people/apps need, rather than something generically useful. That said, the associations involved are somewhat internal, so giving people a way into this is useful… I'll keep mulling it over.

**Comment by grosser** [comment]
I'd think search for multiple tags has a lot of usecase ... and the query is not trivial ... so something nice to have in the library.

**Comment by pat** [comment]
I think what I'd like to see, once #36 is merged in, is `tagged_with` to accept a hash with one of three keys: `:ids`, `:names` or `:tags`, and then an optional key `:match` which defaults to a value of `:any`, but also permits `:all`.

We'd want to keep the existing behaviour, at least for the 0.9.x releases with a deprecation warning.

Does that sound reasonable?

**Comment by grosser** [comment]
means less detection code ... so sounds good to me :)

**Comment by pat** [comment]
Okay, I've got the new `tagged_with` approach set to accept an options hash. Did you want to have a go at adding the `:match` option?

**Comment by benjam1n** [comment]
@pat If implement, would :match allow something like Post.tagged_with(["x","y"], match: true) for showing entries tagged with both x and y? I mean, I've been looking for something like that but can't really figure out a simple solution. Would anyone be kind enough to point me in some direction? Thanks!

**Comment by grosser** [comment]
see my code snippet above

**Comment by grosser** [comment]
not yet implemented / functional ... still on my todo list ... but feel free to open a PR :D

**Comment by benjam1n** [comment]
@grosser Uh oh, you really don't want that, I'm like 50 days into rails... :)
I'm trying to figure out how to apply your snippet to my case but seems more complicated than it should be. What's the "v" variable in your code, an array of tag_id's? Can you just tell me if this could be solvable in like two lines of code or I need to reorg/rewrite my scoping stuff completely? 

```
# post.rb
scope :tag, -> (tag) {
    tags_array = tag.split(' ')
    Post.tagged_with(tags_array)
}

# posts_controller.rb
@posts = @posts.tag(params[:tag]) if params[:tag].present?

```

**Comment by grosser** [comment]
Oh I really want you to make a PR, great learning experience :D

... this should do ... I prefer working with ids when possible to have 1 less join in the mix ...
```Ruby
scope = scope.joins(:tags).where("gutentag_tags.name" => v)
        if v.size == 1
          scope.distinct
        else
          scope.having("count(#{table_name}.id) = #{v.size}").group('#{table_name}.id')
        end
```

**Comment by benjam1n** [comment]
@grosser Michael, thanks a lot. I owe you a beer! :)

For others like me, here's the full solution:

```ruby
scope :tag, -> (tag) do
  tagArray = tag.split(' ')
  scope = Post.joins(:tags).where("gutentag_tags.name": tagArray)
  if tagArray.size == 1
    scope.distinct
  else
    scope.having("count(posts.id) = #{tagArray.size}").group('posts.id')
  end
end
```

**Comment by pat** [comment]
@grosser, @benjam1n: I've just been mucking around in the code, pondering how to make this work. Just for reference, I'm keen to know which databases you're using. PostgreSQL? MySQL? SQLite? Something else?

I'm unsure whether Michael's suggested solution can be applied across all databases, but I'm possibly overcomplicating things.

**Comment by grosser** [comment]
using it on an mysql app atm ... I'd assume it can work on any db ... can
start with a PR that works for mysql and then improve ...

**Comment by pat** [comment]
One difference between the usual MySQL configuration and PostgreSQL (which is what I primarily use) is that a GROUP BY clause in MySQL doesn't require all the unique column to be listed. In PostgreSQL, you'd need to list every single column from the queried model (`Repository` in your initial example, `Post` in @benjam1n's). `GROUP BY posts.*` doesn't cut it, annoyingly.

I'm going to work on expanding the test suite to run across different databases, and then see where we're at after that.

**Comment by benjam1n** [comment]
I'm using postgresql and it seems to work perfectly.
Website | LinkedIn | Twitter

---- On Fri, 02 Jun 2017 08:49:39 +0200 notifications@github.com wrote ----

@grosser, @benjam1n: I've just been mucking around in the code, pondering how to make this work. Just for reference, I'm keen to know which databases you're using. PostgreSQL? MySQL? SQLite? Something else?

I'm unsure whether Michael's suggested solution can be applied across all databases, but I'm possibly overcomplicating things.

**Comment by pat** [comment]
Interesting… I shall investigate further :)

**Comment by pat** [comment]
Okay, turns out you were right and I was wrong. Worked fine in MySQL, PostgreSQL and SQLite. That said, I've ended up taking a bit of a different approach, using a subquery to ensure chaining `count` on the end works as expected.

The related commits, though there's a bunch of refactoring going on:
https://github.com/pat/gutentag/compare/f8cf26940521053bf828dc33d2ac8422f2ec80fe...9c244a36ae43c4ade5957b472c38224486b2d2e9

The behaviour is pretty much equivalent to the following, for `:match => :any` or matching on a single value:

```ruby
subquery = Gutentag::Tagging.select(:taggable_id).
  where(:taggable_type => Post).
  where(:tag_id => [1, 2])
Post.where "id IN (#{subquery.to_sql})"
```

Or, when using `:match => :all`:

```ruby
subquery = Gutentag::Tagging.select(:taggable_id).
  where(:taggable_type => Post).
  where(:tag_id => [1, 2]).
  having("COUNT(*) = 2").
  group(:taggable_id)
Post.where "id IN (#{subquery.to_sql})"
```

**Comment by pat** [comment]
The equivalent of those examples, with the new supported syntax:

```ruby
Post.tagged_with :ids => [1, 2], :match => :any
Post.tagged_with :ids => [1, 2], :match => :all
```

And of course it works if you're using `:ids`, `:tags`, or `:names`.

**Comment by pat** [comment]
One more note: this is now in the freshly released 0.9.0 version! Don't forget to review [the upgrading notes](https://github.com/pat/gutentag#upgrading) when you grab the latest :)

Closing this issue given the requested functionality is implemented. If you find bugs, do let me know (probably best to do so in a new issue).

**Comment by pat** [comment]
@benjam1n: with the new release, your scope can be simplified to the following:

```ruby
scope :tag, -> (tag) do
  tagged_with :names => tag.split(' '), :match => :all
end
```

Also, only because you mentioned you're pretty new to Rails (and Ruby?), even though you're probably already aware of this: the general Ruby convention is to use snake_case for variable names, rather than camelCase (with regards to the `tagArray` variable in your example).

**Comment by benjam1n** [comment]
@pat Very cool. Also, thanks for the advice. :)
